src/main/java/net/consensys/orion/http/handler/privacy/AddToPrivacyGroupHandler.java

```diff
@@ -0,0 +1,164 @@
+import io.vertx.core.buffer.Buffer;
+import io.vertx.ext.web.RoutingContext;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * Add a member to a privacy group
+ */
+public class AddToPrivacyGroupHandler extends PrivacyGroupBaseHandler implements Handler<RoutingContext> {
+
+  private static final Logger log = LogManager.getLogger();
+
+  private final Storage<PrivacyGroupPayload> privacyGroupStorage;
+  private final Storage<QueryPrivacyGroupPayload> queryPrivacyGroupStorage;
+  private final Enclave enclave;
+
+
+  public AddToPrivacyGroupHandler(
+      final Storage<PrivacyGroupPayload> privacyGroupStorage,
+      final Storage<QueryPrivacyGroupPayload> queryPrivacyGroupStorage,
+      final ConcurrentNetworkNodes networkNodes,
+      final Enclave enclave,
+      final Vertx vertx,
+      final Config config) {
+    super(networkNodes, NodeHttpClientBuilder.build(vertx, config, 1500));
+    this.privacyGroupStorage = privacyGroupStorage;
+    this.queryPrivacyGroupStorage = queryPrivacyGroupStorage;
+    this.enclave = enclave;
+  }
+
+  @Override
+  public void handle(final RoutingContext routingContext) {
+    final byte[] request = routingContext.getBody().getBytes();
+    final ModifyPrivacyGroupRequest modifyPrivacyGroupRequest =
+        Serializer.deserialize(JSON, ModifyPrivacyGroupRequest.class, request);
+
+    final AtomicReference<PrivacyGroupPayload> combinedPrivacyGroup = new AtomicReference<>();
+    privacyGroupStorage.get(modifyPrivacyGroupRequest.privacyGroupId()).thenAccept((result) -> {
+      if (result.isEmpty() || result.get().state() == PrivacyGroupPayload.State.DELETED) {
+        routingContext
+            .fail(new OrionException(OrionErrorCode.ENCLAVE_PRIVACY_GROUP_MISSING, "privacy group not found"));
+      } else {
+        final PrivacyGroupPayload oldPrivacyGroupPayload = result.get();
+
+        combinedPrivacyGroup.set(
+            new PrivacyGroupPayload(
+                Stream
+                    .concat(
+                        Arrays.stream(oldPrivacyGroupPayload.addresses()),
+                        Stream.of(modifyPrivacyGroupRequest.address()))
+                    .distinct()
+                    .toArray(String[]::new),
+                oldPrivacyGroupPayload.name(),
+                oldPrivacyGroupPayload.description(),
+                oldPrivacyGroupPayload.state(),
+                oldPrivacyGroupPayload.type(),
+                oldPrivacyGroupPayload.randomSeed()));
+        Stream<Box.PublicKey> combinedAddresses = Arrays
+            .stream(combinedPrivacyGroup.get().addresses())
+            .filter(key -> !key.equals(modifyPrivacyGroupRequest.from()))
+            .distinct()
+            .map(enclave::readKey);
+
+        final var addRequests = sendRequestsToOthers(
+            combinedAddresses,
+            new SetPrivacyGroupRequest(combinedPrivacyGroup.get(), modifyPrivacyGroupRequest.privacyGroupId()),
```

Comment:
Could a byzantine Orion/app construct  a SetPrivacyGroupRequest and send it to a given Orion in order to add itself to the group (or just change the group?). Or is there a layer of encryption/identity which prevents this?

Reply:
There isn't currently. There definitely should be, and I've added an issue for it here:
https://pegasys1.atlassian.net/secure/RapidBoard.jspa?rapidView=57&modal=detail&selectedIssue=OR-316